agentendpoint/agentendpoint.go

```diff
@@ -0,0 +1,160 @@
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+// Package agentendpoint connects to the osconfig agentendpoint api.
+package agentendpoint
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/GoogleCloudPlatform/guest-logging-go/logger"
+	agentendpoint "github.com/GoogleCloudPlatform/osconfig/_internal/gapi-cloud-osconfig-go/cloud.google.com/go/osconfig/agentendpoint/apiv1alpha1"
+	"github.com/GoogleCloudPlatform/osconfig/common"
+	"github.com/GoogleCloudPlatform/osconfig/config"
+	"github.com/GoogleCloudPlatform/osconfig/tasker"
+
+	agentendpointpb "github.com/GoogleCloudPlatform/osconfig/_internal/gapi-cloud-osconfig-go/google.golang.org/genproto/googleapis/cloud/osconfig/agentendpoint/v1alpha1"
+)
+
+// Client is a an agentendpoint client.
+type Client struct {
+	raw *agentendpoint.Client
+}
+
+var client *Client
+
+// GetClient returns the already existing client or creates a new one.
+func GetClient(ctx context.Context) (*Client, error) {
+	if client != nil {
+		return client, nil
+	}
+	c, err := agentendpoint.NewClient(ctx)
+	return &Client{raw: c}, err
+}
+
+// ReportTaskProgress ...
+func (c *Client) ReportTaskProgress(ctx context.Context, req *agentendpointpb.ReportTaskProgressRequest) (res *agentendpointpb.ReportTaskProgressResponse, err error) {
+	logger.Debugf("ReportTaskProgress request:\n%s", common.PrettyFmt(req))
+	if err := retryAPICall(2100*time.Second, "ReportTaskProgress", func() error {
+		res, err = c.raw.ReportTaskProgress(ctx, req)
+		if err != nil {
+			return err
+		}
+		logger.Debugf("Reporting patch details response:\n%s", common.PrettyFmt(res))
+		return nil
+	}); err != nil {
+		return nil, fmt.Errorf("error reporting patch details: %v", err)
+	}
+
+	return
+}
+
+// ReportTaskComplete ...
+func (c *Client) ReportTaskComplete(ctx context.Context, req *agentendpointpb.ReportTaskCompleteRequest) error {
+	logger.Debugf("ReportTaskComplete request:\n%s", common.PrettyFmt(req))
+	if err := retryAPICall(2100*time.Second, "ReportTaskComplete", func() error {
+		res, err := c.raw.ReportTaskComplete(ctx, req)
+		if err != nil {
+			return err
+		}
+		logger.Debugf("Reporting patch details response:\n%s", common.PrettyFmt(res))
+		return nil
+	}); err != nil {
+		return fmt.Errorf("error reporting patch details: %v", err)
+	}
+
+	return nil
+}
+
+func (c *Client) runTask(ctx context.Context) {
+	for {
+		start := &agentendpointpb.ReportTaskStartRequest{
+			Project:    config.ProjectID(),
+			Zone:       config.Zone(),
+			InstanceId: config.ID(),
+		}
+		startRes, err := c.raw.ReportTaskStart(ctx, start)
+		if err != nil {
+			logger.Errorf(err.Error())
+			return
+		}
+
+		task := startRes.GetTask()
+		if task == nil {
+			return
+		}
+
+		switch task.GetTaskType() {
+		case agentendpointpb.TaskType_APPLY_PATCHES:
```

Comment:
We don't need a return statement somewhere inside here to exit runTask?

Reply:
The only time we want to exit is if startRes.GetTask() == nil as that means there is no further task